File: 101/ruby/sort.rb
# frozen_string_literal: true

# Multiple sorting algorithms implementations.
module Sort
  MAX_FIXNUM = (2**(0.size * 8 - 2) - 1)

  # Perform a bubble sort on an array. Sort is performed in place and
  # assumes conflict ordering is not important.
  def bubble_sort(arr)
    arr.each_with_index do |_, i|
      arr.each_with_index do |_, j|
        swap(arr, i, j) if arr[i] < arr[j]
      end
    end
  end

  # Perform an insertion sort on an array. Sort is performed in place and
  # assumes conflict ordering is not important.
  def insertion_sort(arr)
    arr.each_with_index do |_, i|
      j = 0
      while j < i
        if arr[i] < arr[j]
          reposition(arr, i, j)
          break
        end
        j += 1
      end
    end
  end

  # Perform a merge sort on an array. Sort is performed in another array
  # and returned. In and order conflict where items are equal the original
  # ordering remains.
  def merge_sort(arr) # rubocop:disable Metrics/AbcSize
    if arr.size < 2 || (arr.size == 2 && arr[0] < arr[1])
      arr
    elsif arr.size == 2
      [arr[1], arr[0]]
    else
      mid = arr.size / 2
      first = arr.slice(0, mid)
      second = arr.slice(mid, arr.size)
      merge(merge_sort(first), merge_sort(second))
    end
  end

  # Perform a quick sort on an array. Sort is performed in place and
  # assumes conflict ordering is not important.
  def quick_sort(arr, low = 0, high = arr.size - 1)
    return unless low < high

    partition_index = partition(arr, low, high)
    quick_sort(arr, low, partition_index - 1)
    quick_sort(arr, partition_index + 1, high)
  end

  # Perform a selection sort on an array. Sort is performed in place and
  # assumes conflict ordering is not important.
  # rubocop:disable Metrics/MethodLength
  def selection_sort(arr)
    arr.each_with_index do |_, i|
      least = MAX_FIXNUM
      least_index = i
      (i..(arr.length - 1)).each do |j|
        next_val = arr[j]
        if next_val < least
          least = next_val
          least_index = j
        end
      end
      swap(arr, i, least_index) if least_index != i
    end
  end
  # rubocop:enable Metrics/MethodLength

  private

  # Merge two arrays in ascending order.
  # rubocop:disable Metrics/MethodLength
  def merge(first, second)
    i = 0
    j = 0
    result = []
    while i < first.size && j < second.size
      first_val = first[i]
      second_val = second[j]
      if first_val < second_val
        result.push(first_val)
        i += 1
      else
        result.push(second_val)
        j += 1
      end
    end

    result.concat(first.slice(i, first.size))
    result.concat(second.slice(j, second.size))

    result
  end
  # rubocop:enable Metrics/MethodLength

  # Partition a sub array or arr, denoted by the bounding indexes low and high.
  def partition(arr, low, high)
    pivot = arr[high]
    pivot_index = low

    (pivot_index..high).each do |j|
      if arr[j] < pivot
        swap(arr, j, pivot_index)
        pivot_index += 1
      end
    end

    swap(arr, pivot_index, high)

    pivot_index
  end

  # Reposition an element from index `from` to index `to`.
  def reposition(arr, from, to)
    arr.insert(to, arr.delete_at(from))
  end

  # Swap two elements in an array.
  def swap(arr, idx1, idx2)
    tmp = arr[idx1]
    arr[idx1] = arr[idx2]
    arr[idx2] = tmp
  end
end

File: 101/ts/src/sort.ts
/**
 * Perform a Bubble Sort on an array.
 *
 * Sort is performed in place and assumes conflict ordering is
 * not important.
 *
 * @param {Array} array the array to sort.
 */
export const bubbleSort = (arr: Array<number>): void => {
  arr.forEach((_, i) => {
    arr.forEach((_, j) => {
      if (arr[i] < arr[j]) {
        swap(arr, i, j);
      }
    });
  });
};

/**
 * Perform an Insertion Sort on an array.
 *
 * Sort is performed in place and assumes conflict ordering is
 * not important.
 *
 * @param {Array} array the array to sort.
 */
export const insertionSort = (arr: Array<number>): void => {
  arr.forEach((_, i) => {
    let j = 0;

    while (j < i) {
      if (arr[j] > arr[i]) {
        reposition(arr, i, j);
      }
      j++;
    }
  });
};

/**
 * Perform a Merge Sort on an array.
 *
 * Sort is creates a new array in the return val and, where conflict arises,
 * original order is maintained.
 *
 * @param {Array} array the array to sort.
 * @return the sorted array.
 */
export const mergeSort = (arr: Array<number>): Array<number> => {
  if (arr.length < 2 || (arr.length === 2 && arr[1] >= arr[0])) {
    return arr;
  } else if (arr.length === 2) {
    return [arr[1], arr[0]];
  } else {
    const mid = Math.floor(arr.length / 2);
    const first = mergeSort(arr.slice(0, mid));
    const second = mergeSort(arr.slice(mid));
    return merge(first, second);
  }

  function merge(first: Array<number>, second: Array<number>) {
    const result = new Array<number>();
    let i = 0;
    let j = 0;

    while (i < first.length && j < second.length) {
      if (first[i] > second[j]) {
        result.push(second[j]);
        j++;
      } else {
        result.push(first[i]);
        i++;
      }
    }

    return result.concat(first.slice(i), second.slice(j));
  }
};

/**
 * Perform a Quick Sort on an array. Sort is performed in place and assumes
 * conflict order is not important.
 */
export const quickSort = (
  arr: Array<number>,
  low?: number,
  high?: number
): void => {
  const lo = low ?? 0;
  const hi = high ?? arr.length - 1;

  if (lo < hi) {
    const partitionIndex = partition(lo, hi);
    quickSort(arr, lo, partitionIndex - 1);
    quickSort(arr, partitionIndex + 1, hi);
  }

  function partition(low: number, high: number) {
    const pivot = arr[high];
    let pivotIndex = low;

    for (let j = pivotIndex; j < high; j++) {
      if (arr[j] <= pivot) {
        swap(arr, pivotIndex, j);
        pivotIndex++;
      }
    }

    swap(arr, pivotIndex, high);

    return pivotIndex;
  }
};

/**
 * Perform a Selection Sort on an array.
 *
 * Sort is performed in place and assumes conflict ordering is
 * not important.
 *
 * This method could use Math.min.apply and .slice to determine the
 * least value in an array but doesn't to display use of tracking multiple
 * positions.
 *
 * @param {Array} array the array to sort.
 */
export const selectionSort = (arr: Array<number>): void => {
  arr.forEach((_, i) => {
    let least = Number.MAX_SAFE_INTEGER;
    let leastIndex = i;

    for (let j = i; j < arr.length; j++) {
      const next = arr[j];
      if (next < least) {
        least = next;
        leastIndex = j;
      }
    }

    if (leastIndex !== i) {
      swap(arr, i, leastIndex);
    }
  });
};

/**
 * Reposition an element in array.
 *
 * @param {Array} array the array in which to move elements.
 * @param {Number} from the index of the element to move.
 * @param {Number} to the index to move the element to.
 */
const reposition = (array: Array<number>, from: number, to: number) => {
  array.splice(to, 0, array.splice(from, 1)[0]);
};

/**
 * Swap two elements in place, in an array.
 *
 * @param {Array} array they array in which to swap elements.
 * @param {Number} i the index of the first element to swap.
 * @param {Number} j the index of the second element to swap.
 */
const swap = (arr: Array<number>, i: number, j: number) => {
  const tmp = arr[i];
  arr[i] = arr[j];
  arr[j] = tmp;
};
